File: scanner/scanner.py
import attr
from typing import List, Optional, Union
from .exceptions import SyntaxException

from .token_type import TokenType


@attr.s(auto_attribs=True, kw_only=True)
class Scanner:
    """
        <- start = 0
    a
    b
    c
        <- end = 2
    d
    e
    f

    advance() increments "end" by 1
    retreat() decrements "end" by 1

    lexeme() returns the current lexeme
    consume() returns the current lexeme and advances "start" and "end" by its length
    """

    source: str
    start: int = 0
    end: int = 1

    @classmethod
    def tokenize(cls, text: str) -> List["Token"]:
        tokens = []
        scanner = Scanner(source=text)

        single_char_tokens = {
            "(": TokenType.LEFT_PAREN,
            ")": TokenType.RIGHT_PAREN,
            "{": TokenType.LEFT_BRACE,
            "}": TokenType.RIGHT_BRACE,
            ",": TokenType.COMMA,
            ".": TokenType.DOT,
            "-": TokenType.MINUS,
            "+": TokenType.PLUS,
            ";": TokenType.SEMICOLON,
            "*": TokenType.STAR,
            "/": TokenType.SLASH,
        }

        line = 1
        while not scanner.at_end():
            token_type = None
            literal: Optional[Union[str, int, float]] = None

            if scanner.lexeme() == "\n":
                line += 1

            elif scanner.lexeme() in ("\r", "\t", " "):
                pass

            elif scanner.lexeme() == "#":
                while scanner.lookahead() != "\n":
                    scanner.advance()

            elif scanner.lexeme() in single_char_tokens:
                token_type = single_char_tokens[scanner.lexeme()]

            elif scanner.lexeme() == "!":
                if scanner.match("="):
                    token_type = TokenType.BANG_EQUAL
                else:
                    token_type = TokenType.BANG

            elif scanner.lexeme() == "=":
                if scanner.match("="):
                    token_type = TokenType.EQUAL_EQUAL
                else:
                    token_type = TokenType.EQUAL

            elif scanner.lexeme() == ">":
                if scanner.match("="):
                    token_type = TokenType.GREATER_EQUAL
                else:
                    token_type = TokenType.GREATER

            elif scanner.lexeme() == "<":
                if scanner.match("="):
                    token_type = TokenType.LESS_EQUAL
                else:
                    token_type = TokenType.LESS

            elif scanner.lexeme() in ('"', "'"):
                quote = scanner.lexeme()
                token_type = TokenType.STRING
                while not scanner.match(quote):
                    if scanner.at_end() or scanner.lookahead() == "\n":
                        raise SyntaxException(f"Expected {quote}")

                    scanner.advance()
                    continue
                literal = scanner.lexeme()[1:-1]

            elif cls._is_digit(scanner.lexeme()):
                token_type = TokenType.INTEGER
                while cls._is_digit(scanner.lookahead()):
                    scanner.advance()

                literal = int(scanner.lexeme())

                if scanner.match("."):
                    token_type = TokenType.FLOAT
                    while cls._is_digit(scanner.lookahead()):
                        scanner.advance()
                    literal = float(scanner.lexeme())

            elif cls._is_alpha(scanner.lexeme()):
                token_type = TokenType.IDENTIFIER

                while cls._is_alpha_numeric(scanner.lookahead()):
                    scanner.advance()

                keyword = TokenType.match_keyword(scanner.lexeme())
                if keyword:
                    token_type = keyword

            lexeme = scanner.consume()

            if token_type:
                tokens.append(
                    Token(
                        token_type=token_type, lexeme=lexeme, line=line, literal=literal
                    )
                )
            elif lexeme[0] in ("\n", "\t", "\r", "#", " "):
                pass
            else:
                raise SyntaxException(f"Syntax Error on line {line}: '{lexeme}'")

        tokens.append(Token(token_type=TokenType.EOF, lexeme="", line=line))

        return tokens

    def advance(self, amount: int = 1):
        self.end += amount

    def consume(self) -> str:
        lexeme = self.lexeme()
        self.start, self.end = self.end, self.end + 1
        return lexeme

    def lexeme(self) -> str:
        return self.source[self.start : self.end]

    def lookahead(self, count=1) -> str:
        return self.source[self.end : (self.end + count)]

    def match(self, target: str) -> bool:
        if self.source[self.end : (self.end + len(target))] != target:
            return False

        self.advance(len(target))
        return True

    def retreat(self):
        self.end -= 1

    def at_end(self) -> bool:
        return self.end > len(self.source)

    @classmethod
    def _is_digit(cls, char: str) -> bool:
        return "0" <= char <= "9"

    @classmethod
    def _is_alpha(cls, char: str) -> bool:
        return ("a" <= char <= "z") or ("A" <= char <= "Z") or char == "_"

    @classmethod
    def _is_alpha_numeric(cls, char: str) -> bool:
        return cls._is_alpha(char) or cls._is_digit(char)

File: c-sharp/scanner.cs
internal class Scanner {
    /**
     * Class for scanning lox code for tokens.
     */

    public static Token[] getTokens(string code) {
        /** Use two-pointers (start and end) to scan the code for tokens.
         *
         *   var mystring = "bar";
         *       ^       ^
         *       s       e
         *       t       n
         *       a       d
         *       r
         *       t
         */

        Scanner scanner = new Scanner(code);
        List<Token> tokens = new List<Token>();
        int line = 1;

        while (true) {
            switch(scanner.lexeme()) {
                case "\n":
                    line++;
                    break;
                case "/":
                    if (scanner.match("/")) {
                        // A comment goes until the end of the line.
                        while (scanner.lookahead() != "\n" && !scanner.atEnd()) scanner.advance();
                    } else {
                        tokens.Add(new Token(Token.Types.SLASH, scanner.lexeme(), null, line));
                    }
                    break;
                case "(":
                case ")":
                case "{":
                case "}":
                case ",":
                case ".":
                case "-":
                case "+":
                case ";":
                case "*":
                    tokens.Add(Token.matchingSingleCharToken(scanner.lexeme(), line));
                    break;
                case "!":
                    tokens.Add(new Token(scanner.match("=") ? Token.Types.BANG_EQUAL: Token.Types.BANG, scanner.lexeme(), null, line));
                    break;
                case "=":
                    tokens.Add(new Token((scanner.match("=")) ? Token.Types.EQUAL_EQUAL: Token.Types.EQUAL, scanner.lexeme(), null, line));
                    break;
                case ">":
                    tokens.Add(new Token(scanner.match("=") ? Token.Types.GREATER_EQUAL: Token.Types.GREATER, scanner.lexeme(), null, line));
                    break;
                case "<":
                    tokens.Add(new Token(scanner.match("=") ? Token.Types.LESS_EQUAL : Token.Types.LESS, scanner.lexeme(), null, line));
                    break;
                case "'":
                case "\"":
                    var quote = scanner.lexeme();
                    var token_type = Token.Types.STRING;

                    while (scanner.lookahead() != quote && !scanner.atEnd()) {
                        if (scanner.lookahead() == "\n") line++;
                        scanner.advance();
                    }
                    break;
                default:
                    if (scanner.isDigit(scanner.lexeme())) {
                        while (scanner.isDigit(scanner.lexeme())) scanner.advance();

                        // Look for a fractional part.
                        if (scanner.lookahead() == "." && scanner.isDigit(scanner.lookahead(2))) {
                            // Consume the "."
                            scanner.advance();

                            while (scanner.isDigit(scanner.lexeme())) scanner.advance();
                        }

                        tokens.Add(new Token(Token.Types.NUMBER, scanner.lexeme(), Double.Parse(scanner.lexeme()), line));
                    }

                    else if (scanner.isAlpha(scanner.lexeme())) {
                        while (scanner.isAlphaNumeric(scanner.lexeme())) scanner.advance();

                        // See if the identifier is a keyword.
                        var word = scanner.lexeme();
                        var type = Token.Types.IDENTIFIER;

                        if (Token.keywords.ContainsKey(word)) {
                            type = Token.keywords[word];
                        }

                        tokens.Add(new Token(type, word, null, line));
                    }
                    break;
            }
        }

        tokens.Add(new Token(Token.Types.EOF, "", null, line));

        return tokens.ToArray();
    }

    private string code;
    private int start = 0;
    private int end = 1;

    private Scanner(string code) {
        this.code = code;
    }


    private void advance(int amount = 1) {
        this.end += amount;
    }

    private string consume() {
        var lexeme = this.lexeme();
        (this.start, this.end) = (this.end, this.end + 1);
        return lexeme;
    }

    private string lexeme() {
        return this.code.Substring(this.start, this.end);
    }

    private string lookahead(int amount = 1) {
        return this.code.Substring(this.end, this.end + amount);
    }

    private bool match(string target) {
        if (this.lookahead(target.Length) == target) {
            this.advance();
            return true;
        }
        return false;
    }

    private void retreat() {
        this.end--;
    }

    private bool atEnd() {
        return this.end > this.code.Length;
    }

    private bool isDigit(string c) {
        return "0123456789".Contains(c[0]);
    }

    private bool isAlpha(string c) {
        return "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_".Contains(c[0]);
    }

    private bool isAlphaNumeric(string c) {
        return this.isAlpha(c) || this.isDigit(c);
    }
}
